## Goals for the next meeting (Fri, 10/2)

+ impute missing values using KNN
+ complete PCA for credit and medical payment
+ potentially fit the baseline model!!


Updated: Wed, 9/30

In [1]:
import pandas as pd
import numpy as np

from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
from sklearn.decomposition import TruncatedSVD

## Import variable files

In [2]:
f = open('./health-PCA.txt', 'r')
health_pca_var_names = f.read().split("\n")
f = open('./credit-PCA.txt', 'r')
credit_pca_var_names = f.read().split("\n")
f = open('./medical-payment-PCA.txt', 'r')
med_pay_pca_var_names = f.read().split("\n")
f = open('./other-variables.txt', 'r')
var_names = f.read().split("\n")

In [3]:
health_pca_var_names = [var for var in pd.read_fwf('./health-PCA.txt',header=None)[0]]
credit_pca_var_names = [var for var in pd.read_fwf('./credit-PCA.txt',header=None)[0]]
med_pay_pca_var_names = [var for var in pd.read_fwf('./medical-payment-PCA.txt',header=None)[0]]
var_names = [var for var in pd.read_fwf('./other-variables.txt',header=None)[0]]

## Read datasets

Divided the original dataset into 4 smaller subsets.

+ `health_df`: dataset for health condition-related variables (no missing value - PCA is ok)
+ `credit_df`: dataset for financial information (need to impute missing values)
+ `med_pay_df`: dataset for medical payment related variables (need to impute missing values)
+  `df`: dataset for variables that do not need PCA (need to impute missing values)

In [4]:
health_df = pd.read_csv('./2020_Competition_Training.csv', usecols=health_pca_var_names,low_memory = False)
credit_df = pd.read_csv('./2020_Competition_Training.csv', usecols=credit_pca_var_names,low_memory = False)
med_pay_df = pd.read_csv('./2020_Competition_Training.csv', usecols=med_pay_pca_var_names,low_memory = False)
df = pd.read_csv('./2020_Competition_Training.csv', usecols=var_names,low_memory = False)

## PCA for `health_df`

In [5]:
pca_health = PCA(n_components = 80)
X = pca_health.fit_transform(health_df)

In [6]:
health_df.shape

(69572, 224)

In [7]:
sum(pca_health.explained_variance_ratio_)

0.8961624187712404

Reduced dimentions from 224 to 80 and retained ~ 90% variance

## The rest of the datasets

+ The remaining datasets contains NaN values -- PCA or other algorithms cannot be done. 
+ Need to impute missing values.
+ Current idea of how to impute: KNN

See this [reference](https://towardsdatascience.com/xgboost-is-not-black-magic-56ca013144b4) and [this](https://scikit-learn.org/stable/modules/generated/sklearn.impute.KNNImputer.html) for implementation.

In [8]:
credit_df.shape

(69572, 20)

In [9]:
# All rows with missing values in credit_df
credit_missing = credit_df[credit_df.isnull().any(axis=1)]
credit_missing

credit_bal_autobank  credit_bal_autofinance  \
22                     NaN                     NaN   
546                    NaN                     NaN   
554                    NaN                     NaN   
622                    NaN                     NaN   
625                    NaN                     NaN   
...                    ...                     ...   
67788                  NaN                     NaN   
67932                  NaN                     NaN   
68214                  NaN                     NaN   
68783                  NaN                     NaN   
69097                  NaN                     NaN   

       credit_bal_bankcard_severederog  credit_bal_consumerfinance  \
22                                 NaN                         NaN   
546                                NaN                         NaN   
554                                NaN                         NaN   
622                                NaN                         NaN   
625                                NaN                         NaN   
...                                ...                         ...   
67788                              NaN                         NaN   
67932                              NaN                         NaN   
68214                              NaN                         NaN   
68783                              NaN                         NaN   
69097                              NaN                         NaN   

       credit_bal_heloc_60dpd  credit_bal_heloc_severederog  \
22                        NaN                           NaN   
546                       NaN                           NaN   
554                       NaN                           NaN   
622                       NaN                           NaN   
625                       NaN                           NaN   
...                       ...                           ...   
67788                     NaN                           NaN   
67932                     NaN                           NaN   
68214                     NaN                           NaN   
68783                     NaN                           NaN   
69097                     NaN                           NaN   

       credit_bal_mtg_90to119dpd  credit_bal_mtg_bankruptcy  \
22                           NaN                        NaN   
546                          NaN                        NaN   
554                          NaN                        NaN   
622                          NaN                        NaN   
625                          NaN                        NaN   
...                          ...                        ...   
67788                        NaN                        NaN   
67932                        NaN                        NaN   
68214                        NaN                        NaN   
68783                        NaN                        NaN   
69097                        NaN                        NaN   

       credit_bal_mtg_severederog  credit_bal_mtgcredit_new  \
22                            NaN                       NaN   
546                           NaN                       NaN   
554                           NaN                       NaN   
622                           NaN                       NaN   
625                           NaN                       NaN   
...                           ...                       ...   
67788                         NaN                       NaN   
67932                         NaN                       NaN   
68214                         NaN                       NaN   
68783                         NaN                       NaN   
69097                         NaN                       NaN   

       credit_bal_studentloan_60dpd  credit_bal_totalallcredit_60dpd  \
22                              NaN                              NaN   
546                             NaN                              NaN   
554                           

In [10]:
med_pay_df.shape

(69572, 13)

In [11]:
# NA in med_pay_df
med_pay_missing = med_pay_df[med_pay_df.isnull().any(axis=1)]
med_pay_missing

cms_ma_risk_score_nbr  cms_partd_ra_factor_amt  \
0                        NaN                      NaN   
66                       NaN                      NaN   
121                      NaN                      NaN   
123                      NaN                      NaN   
129                      NaN                      NaN   
...                      ...                      ...   
69480                    NaN                      NaN   
69520                    NaN                      NaN   
69542                    NaN                      NaN   
69543                    NaN                      NaN   
69565                    NaN                      NaN   

       cms_risk_adj_payment_rate_a_amt  cms_risk_adj_payment_rate_b_amt  \
0                                  NaN                              NaN   
66                                 NaN                              NaN   
121                                NaN                              NaN   
123                                NaN                              NaN   
129                                NaN                              NaN   
...                                ...                              ...   
69480                              NaN                              NaN   
69520                              NaN                              NaN   
69542                              NaN                              NaN   
69543                              NaN                              NaN   
69565                              NaN                              NaN   

       cms_risk_adjustment_factor_a_amt  cms_rx_risk_score_nbr  \
0                                   NaN                    NaN   
66                                  NaN                    NaN   
121                                 NaN                    NaN   
123                                 NaN                    NaN   
129                                 NaN                    NaN   
...                                 ...                    ...   
69480                               NaN                    NaN   
69520                               NaN                    NaN   
69542                               NaN                    NaN   
69543                               NaN                    NaN   
69565                               NaN                    NaN   

       cms_tot_ma_payment_amt  cms_tot_partd_payment_amt  rev_cms_clinic_ind  \
0                         NaN                        NaN                   0   
66                        NaN                        NaN                   0   
121                       NaN                        NaN                   0   
123                       NaN                        NaN                   0   
129                       NaN                        NaN                   0   
...                       ...                        ...                 ...   
69480                     NaN                        NaN                   0   
69520                     NaN                        NaN                   0   
69542                     NaN                        NaN                   0   
69543                     NaN                        NaN                   0   
69565                     NaN                        NaN                   0   

       rev_cms_er_ind  rev_cms_icu_ind  rev_cms_lab_ind  rev_cms_phar_ind  
0                   0                0                0                 0  
66                  0                0                0                 0  
121                 0                0                0                 0  
123                 0                0                0                 0  
129                 1                1                1                 1  
...               ...              ...              ...               ...  
69480               1                0                1                 1  
69520               0                0                0    

In [12]:
df[df.isnull().any(axis = 1)].shape

(21582, 40)

In [13]:
# NaN in df
df[df.isnull().any(axis = 1)]

sex_cd  est_age  smoker_current_ind  smoker_former_ind lang_spoken_cd  \
0          F       62                   1                  0            ENG   
3          M       75                   0                  0            ENG   
8          F       56                   0                  0            ENG   
11         F       67                   0                  0            SPA   
15         M       60                   0                  1            ENG   
...      ...      ...                 ...                ...            ...   
69564      F       70                   0                  0            ENG   
69565      F       18                   0                  0            ENG   
69567      F       72                   1                  0            ENG   
69568      M       75                   0                  0            ENG   
69571      M       60                   0                  0            ENG   

      mabh_seg  cci_score  dcsi_score  fci_score  hcc_weighted_sum  ...  \
0          UNK        3.0           1          2                 0  ...   
3           H6        3.0           0          2                 0  ...   
8           H3        6.0           3          1                 0  ...   
11         UNK        6.0           1          4                 0  ...   
15          H2        4.0           2          2                 3  ...   
...        ...        ...         ...        ...               ...  ...   
69564      UNK        4.0           3          5                 4  ...   
69565      UNK        0.0           0          0                 0  ...   
69567       H7        5.0           2          2                 2  ...   
69568       C4        9.0           3          2                14  ...   
69571      UNK        2.0           3          2                 0  ...   

       phy_em_pe_ind  phy_em_pi_ind  phy_em_px_ind prov_fb_ind prov_pcp_ind  \
0                  0              0              1           1            1   
3                  0              0              1           0            1   
8                  1              0              1           1            1   
11                 0              0              1           1            1   
15                 0              0              1           0            1   
...              ...            ...            ...         ...          ...   
69564              0              0              1           1            1   
69565              0              0              0           0            0   
69567              0              0              1           1            1   
69568              0              0              1           1            1   
69571              0              0              1           1            1   

       prov_sp_ind  rucc_category rx_overall_pmpm_ct  \
0                1        1-Metro           0.084756   
3                1     7-Nonmetro           2.333333   
8                1        1-Metro           3.433116   
11               1        1-Metro          12.909578   
15               1        2-Metro           2.299222   
...            ...            ...                ...   
69564            1        2-Metro           6.876191   
69565            0     6-Nonmetro           0.384005   
69567            1        1-Metro           3.916667   
69568            1        1-Metro           4.872273   
69571            1        2-Metro           6.058179   

      total_outpatient_visit_ct_pmpm  total_physician_office_visit_ct_pmpm  
0                           0.000000                              0.250900  
3                           0.083333                              0.250000  
8                           0.250324                              0.610793  
11                          0.831499                              2.838363  
15                          0.125749                              0.581078  
...                              ...                                   .

In [14]:
df_missing=df.isnull().any()

In [15]:
df_missing[df_missing == True].index

Index(['cms_ra_factor_type_cd', 'cons_cmys', 'cons_hcaccprf_h',
       'cons_hcaccprf_p', 'cons_hhcomp', 'cons_homstat', 'cons_n2029_y',
       'cons_n2mob', 'cons_n2pbl', 'cons_n2pmv', 'cons_n65p_y',
       'cons_retail_buyer', 'hedis_dia_hba1c_ge9', 'hedis_dia_hba1c_test',
       'hedis_dia_ma_nephr'],
      dtype='object')

In [16]:
df[df.isnull().any(axis = 1)][df_missing[df_missing == True].index]

cms_ra_factor_type_cd cons_cmys  cons_hcaccprf_h  cons_hcaccprf_p  \
0                       NaN       NaN              NaN              NaN   
3                        CN       NaN              NaN              NaN   
8                         E       NaN              NaN              NaN   
11                       CP       NaN              NaN              NaN   
15                       CP       NaN              NaN              NaN   
...                     ...       ...              ...              ...   
69564                    CN       NaN              NaN              NaN   
69565                   NaN       NaN              NaN              NaN   
69567                    CN       NaN              NaN              NaN   
69568                    CF       NaN              NaN              NaN   
69571                     E       NaN              NaN              NaN   

      cons_hhcomp cons_homstat  cons_n2029_y  cons_n2mob  cons_n2pbl  \
0             NaN          NaN           NaN         NaN         NaN   
3             NaN          NaN           NaN         NaN         NaN   
8             NaN          NaN           NaN         NaN         NaN   
11            NaN          NaN           NaN         NaN         NaN   
15            NaN          NaN           NaN         NaN         NaN   
...           ...          ...           ...         ...         ...   
69564         NaN          NaN           NaN         NaN         NaN   
69565         NaN          NaN           NaN         NaN         NaN   
69567         NaN          NaN           NaN         NaN         NaN   
69568         NaN          NaN           NaN         NaN         NaN   
69571         NaN          NaN           NaN         NaN         NaN   

       cons_n2pmv  cons_n65p_y  cons_retail_buyer hedis_dia_hba1c_ge9  \
0             NaN          NaN                NaN                   N   
3             NaN          NaN                NaN                   N   
8             NaN          NaN                NaN                   N   
11            NaN          NaN                NaN                   N   
15            NaN          NaN                NaN                   N   
...           ...          ...                ...                 ...   
69564         NaN          NaN                NaN                   N   
69565         NaN          NaN                NaN                 NaN   
69567         NaN          NaN                NaN                   N   
69568         NaN          NaN                NaN                   N   
69571         NaN          NaN                NaN                   N   

      hedis_dia_hba1c_test hedis_dia_ma_nephr  
0                        N                  Y  
3                        Y                  Y  
8                        Y                  Y  
11                       Y                  Y  
15                       N                  Y  
...                    ...                ...  
69564                    Y                  N  
69565                  NaN                NaN  
69567                    Y                  Y  
69568                    Y                  Y  
69571                    Y                  Y  

[21582 rows x 15 columns]

In [17]:
from sklearn.impute import KNNImputer

In [18]:
credit_df.sample(10)

credit_bal_autobank  credit_bal_autofinance  \
44308          5472.927989             5130.862707   
3213           4317.461317             2517.959277   
2878           5957.968747             6726.431914   
30232          5982.209071             4503.288370   
59381          6310.586918             4049.737153   
16171          4100.569828             5659.898125   
61367          3111.447710             4965.983516   
11663          5286.393968             4381.416478   
23490          3356.511247             7960.167146   
43821          3830.573113             3926.501377   

       credit_bal_bankcard_severederog  credit_bal_consumerfinance  \
44308                       349.418943                  819.436502   
3213                        399.555783                  528.374975   
2878                        511.416480                  611.364441   
30232                       243.261162                  649.971825   
59381                       317.651916                  641.478406   
16171                       275.015570                  673.715987   
61367                       211.675891                  843.434061   
11663                       255.858632                  747.295909   
23490                       340.048215                  833.227517   
43821                       278.205913                 1116.602154   

       credit_bal_heloc_60dpd  credit_bal_heloc_severederog  \
44308               24.994429                     39.723341   
3213                24.936778                      2.866883   
2878                80.984915                     18.240953   
30232               14.650237                      6.879945   
59381              148.876428                    150.106468   
16171               37.144860                      2.289875   
61367                6.189463                      7.683590   
11663               52.984190                     38.946131   
23490               85.934544                      5.874713   
43821                0.259484                      0.000000   

       credit_bal_mtg_90to119dpd  credit_bal_mtg_bankruptcy  \
44308                  10.526386                 345.614037   
3213                   70.895779                  76.215218   
2878                  489.259079                 169.383899   
30232                 253.761982                 478.204677   
59381                  96.407884                 148.816484   
16171                  61.576647                 168.798679   
61367                 184.223217                  79.270946   
11663                  35.391060                 115.346888   
23490                  73.991865                 292.328307   
43821                  47.838432                  64.142995   

       credit_bal_mtg_severederog  credit_bal_mtgcredit_new  \
44308                  266.596157               9125.147291   
3213                    32.787334               8041.671225   
2878                   101.712643              17596.173660   
30232                   37.117296               7161.768654   
59381                  233.757731               7754.004725   
16171                   36.274208               7110.213798   
61367                   40.064922               7691.393532   
11663                  133.124628               6297.306763   
23490                   40.267374               7820.891540   
43821                   23.186035               2134.358211   

       credit_bal_studentloan_60dpd  credit_bal_totalallcredit_60dpd  \
44308                    566.924916                      3553.992120   
3213                    1893.741768                      6748.269785   
2878                    1364.812301                      6915.536698   
30232                   2335.465042                      5267.073285   
59381                   1037.202387                      3553.237995   
16171                   1547.763639                      3844.373003   
61367                   1494.407553                    

In [19]:
imputer = KNNImputer(n_neighbors=5)
df1 = pd.DataFrame(imputer.fit_transform(credit_missing))
df1

0            1         2
0    404.451994  3407.307692  4.188668
1    404.451994  3407.307692  4.188668
2    404.451994  3407.307692  4.188668
3    404.451994  3407.307692  4.188668
4    404.451994  3407.307692  4.188668
..          ...          ...       ...
226  404.451994  3407.307692  4.188668
227  404.451994  3407.307692  4.188668
228  404.451994  3407.307692  4.188668
229  404.451994  3407.307692  4.188668
230  404.451994  3407.307692  4.188668

[231 rows x 3 columns]

In [20]:
imputer = KNNImputer(n_neighbors=5)
df2 = pd.DataFrame(imputer.fit_transform(med_pay_missing))
df2

0       1        2        3       4       5         6        7   \
0     1.1782  2.4998  347.742  430.622  1.1116  0.9634   705.892  104.002   
1     1.1782  2.4998  347.742  430.622  1.1116  0.9634   705.892  104.002   
2     1.1782  2.4998  347.742  430.622  1.1116  0.9634   705.892  104.002   
3     1.1782  2.4998  347.742  430.622  1.1116  0.9634   705.892  104.002   
4     2.3188  1.1312  503.842  747.566  2.2482  1.3442  1507.278  104.148   
...      ...     ...      ...      ...     ...     ...       ...      ...   
3879  1.7338  1.3750  524.030  610.704  1.1908  0.8642  1299.200   74.754   
3880  1.1782  2.4998  347.742  430.622  1.1116  0.9634   705.892  104.002   
3881  1.7338  1.3750  524.030  610.704  1.1908  0.8642  1299.200   74.754   
3882  0.9192  0.5258  367.798  496.956  1.1216  0.7062   882.444   96.060   
3883  1.0780  1.9670  365.968  519.646  1.0540  1.0026   802.350  128.562   

       8    9    10   11   12  
0     0.0  0.0  0.0  0.0  0.0  
1     0.0  0.0  0.0  0.0  0.0  
2     0.0  0.0  0.0  0.0  0.0  
3     0.0  0.0  0.0  0.0  0.0  
4     0.0  1.0  1.0  1.0  1.0  
...   ...  ...  ...  ...  ...  
3879  0.0  1.0  0.0  1.0  1.0  
3880  0.0  0.0  0.0  0.0  0.0  
3881  0.0  1.0  0.0  1.0  1.0  
3882  0.0  0.0  0.0  1.0  0.0  
3883  0.0  0.0  0.0  0.0  0.0  

[3884 rows x 13 columns]